### PR TITLE
platform: Configuration management MVP

### LINKED ISSUE #49: Anchor Platform Config Interface
Create the anchor platform configuration management interface so development of modules can begin using without configuration management backend implementations.

### PR DESCRIPTION
**Pull request by lijamie98**
closes #49 

<details>
  <summary>PR Checklist</summary>

### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

MVP implementation configuration management. 

### Why

Configuration management for sep, payment-gateway, and data-access.

### Known limitations

Work in progress. Rough edges need smoothing.

### CODE REVIEW

File: docs/anchor-platform.md [hunk 1]
```diff
@@ -1,4 +1,53 @@
 # Anchor Platform Spring Boot Application
 
 ## To Run
-./gradlew bootRun
+```shell
+./gradlew bootRun -PjvmArgs="-Dstellar.anchor.config=path-to-yaml"
```
Comment (marcelosalloum) [inline docs/anchor-platform.md:5]:
```suggestion
./gradlew bootRun -PjvmArgs="-Dstellar.anchor.config=[path-to-yaml]" 
```

File: docs/anchor-platform.md [hunk 2]
```diff
@@ -1,4 +1,53 @@
...
+
+```shell
+./gradlew bootRun -PjvmArgs="-Dstellar.anchor.config=[path-to-yaml]" 
+```
+
+### 2. System Environment Variable `STELLAR_ANCHOR_CONFIG`
+The path of the `yaml` file can be specified by the system environment variable `STELLAR_ANCHOR_CONFIG`.
+
+```shell
+export STELLAR_ANCHOR_CONFIG=[file-to-yaml]
```
Comment (marcelosalloum) [inline docs/anchor-platform.md:28]:
```suggestion
export STELLAR_ANCHOR_CONFIG=[path-to-yaml]
```

File: docs/anchor-platform.md (hunk 1 above)
Comment (marcelosalloum) [inline docs/anchor-platform.md:5]: Is this path absolute or relative?

File: docs/anchor-platform.md [hunk 3]
```diff
@@ -1,4 +1,53 @@
...
+### Environment variables
+
+In the scenarios where a CMS or Consul is available when there are few variables to pass, it is practical to pass through environment variables.
+
+## Example YAML file
+[An example of the yaml file](../platform/example.anchor-config.yaml).
+
+
+
+
```
Comment (marcelosalloum) [inline docs/anchor-platform.md:53]: Deleting unnecessary lines:
```suggestion
```

File: payment-circle/src/main/java/org/stellar/anchor/paymentservice/circle/config/StellarxPaymentConfig.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.stellar.anchor.paymentservice.circle.config;
+
+/**
+ * This is a placeholder class. TODO: move to payment-stellarx subproject under
+ * org.stellar.anchor.paymentservice.circle.config package.
+ */
+public interface StellarxPaymentConfig {
```
Comment (marcelosalloum) [inline payment-circle/src/main/java/org/stellar/anchor/paymentservice/circle/config/StellarxPaymentConfig.java:7]: Same here: Stellarx should be `Stellar`. StellarX is this other product that uses Stellar: https://www.stellarx.com/

File: platform/example.anchor-config.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
+stellar:
+  anchor:
+    # Configure the application from this file after it is loaded in memory
+    # If `config` is in-memory, the yaml file contains all settings for the server.
+    config: in-memory
+    #
+    # Configures how the applications read the
+    #
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:8]: This text is confusing.

File: platform/example.anchor-config.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      type: logging-logback
+      settings: logging-logback-settings
+
+#
+# Application settings
+#
+app-config:
+  # general
+  app:
+    stellarNetworkPassPhrase: Test SDF Network ; September 2015
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:25]: nit: `passphrase` is usually a single word, like in this json response: https://horizon-testnet.stellar.org/

```suggestion
    stellarNetworkPassphrase: Test SDF Network ; September 2015
```

File: platform/example.anchor-config.yaml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  sep1:
+    enabled: true
+    stellarFile: sep1/stellar-wks.toml
+
+  # sep-10
+  sep10:
+    enabled: true
+    homeDomain: localhost:8080
+    clientAttributionRequired: false
+    clientAttributionAllowList: lobstr.co,preview.lobstr.co
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:42]: In a real case scenario, should we allow all kinds of domains? If so, how to do that?

File: platform/example.anchor-config.yaml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+#
+# Application settings
+#
+app-config:
+  # general
+  app:
+    stellarNetworkPassPhrase: Test SDF Network ; September 2015
+    hostUrl: http://localhost:8080
+    languages: en
+    horizonURI: https://horizon-testnet.stellar.org
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:28]: In most places within this project, URI and URL are spelt in camelcase: `horizonUri`. I think we should be consistent here.

File: platform/example.anchor-config.yaml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  payment-gateway:
+    #
+    # Payment Circle configurations
+    #
+    circle:
+      name: "circle_payment"
+      enabled: true
+      circleUrl: http://payment.circle.com
+      secretKey: This-is-a-secret-key
+      horizonUrl: http://horizon.stellar.org
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:61]: Can this inherit from the app-config? Not sure we could have variables in a yaml file.

Also, we should default to testnet:
```suggestion
      horizonUrl: https://horizon-testnet.stellar.org
```

File: platform/example.anchor-config.yaml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    interactiveUrl: http://localhost:8080/ref/sep24/interactive
+
+  payment-gateway:
+    #
+    # Payment Circle configurations
+    #
+    circle:
+      name: "circle_payment"
+      enabled: true
+      circleUrl: http://payment.circle.com
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:59]: We should default to the sandbox URL here:

```suggestion
      circleUrl: https://api-sandbox.circle.com
```

File: platform/example.anchor-config.yaml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  sep24:
+    interactiveJwtExpiration: 300
+    interactiveUrl: http://localhost:8080/ref/sep24/interactive
+
+  payment-gateway:
+    #
+    # Payment Circle configurations
+    #
+    circle:
+      name: "circle_payment"
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:57]: `circle` should be enough here
```suggestion
      name: "circle"
```

File: platform/example.anchor-config.yaml [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      horizonUrl: http://horizon.stellar.org
+      stellarNetwork: TESTNET
+
+    #
+    # Payment StellarX configurations
+    #
+    stellarx:
+      enabled: false
+      name: "stellarx_payment"
+      horizon_url: http://horizon.stellar.org
```
Comment (marcelosalloum) [inline platform/example.anchor-config.yaml:70]: Suggested changes:
* Simplify the name
* Rename the variable for consistency
* Default to testnet

```suggestion
      name: "stellar"
      horizonUrl: https://horizon-testnet.stellar.org
```

File: platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  @Override
+  public void initialize(@NotNull ConfigurableApplicationContext applicationContext) {
+    loadConfigYaml("default", new ClassPathResource("default-config.yaml"));
+
+    // Read from $USER_HOME/.anchor/anchor-config.yaml
+    File yamlFile = new File(System.getProperty("user.home"), "./anchor/anchor-config.yaml");
+    if (yamlFile.exists()) {
+      loadConfigYaml("yaml", new FileSystemResource(yamlFile));
+      return;
+    }
```
Comment (marcelosalloum) [inline platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java:32]: Wait, shouldn't this be the last option according to the documentation?

File: platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    if (yamlFilePath != null) {
+      loadConfigYaml("yaml", new FileSystemResource(yamlFilePath));
+      return;
+    }
+
+    // Read from Java VM
+    yamlFilePath = System.getProperty("stellar.anchor.config");
+    if (yamlFilePath != null) {
+      loadConfigYaml("yaml", new FileSystemResource(yamlFilePath));
+    }
```
Comment (marcelosalloum) [inline platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java:45]: And I think the documentation states this is the first option.

**Review by marcelosalloum — COMMENTED** [review]
Looking great! I've left a few comments on small improvements, fixes and typos ⬇️

**Review by marcelosalloum — APPROVED** [review]
Approving because none of those comments is critical. @lijamie98, plz feel free to merge this code once the fixes are in place.

File: config-spring-property/src/main/java/org/stellar/anchor/server/config/payment/PropertyStellarxPaymentConfig.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package org.stellar.anchor.server.config.payment;
+
+import lombok.Data;
+import org.stellar.anchor.paymentservice.circle.config.StellarxPaymentConfig;
+
+@Data
+public class PropertyStellarxPaymentConfig implements StellarxPaymentConfig {
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply config-spring-property/src/main/java/org/stellar/anchor/server/config/payment/PropertyStellarxPaymentConfig.java:7]: Changed.

File: core/src/main/java/org/stellar/anchor/paymentservice/PaymentGateway.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+  public static class Builder {
+    HashMap<String, PaymentService> map = new HashMap<>();
+
+    public PaymentGateway build() {
+      return new PaymentGateway(map);
+    }
+
+    public Builder add(PaymentService service) {
+      if (map.get(service.getName()) != null) {
+        throw new RuntimeException("The serivce [" + service.getName() + "] already exists.");
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/paymentservice/PaymentGateway.java:32]: I was thinking about this as well. If that's the case, I will create another PR for multi-instance payment services.

File: docs/anchor-platform.md (hunk 1 above)
Comment (lijamie98) [reply docs/anchor-platform.md:5]: Thank you!

File: platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java (hunk 13 above)
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/server/configurator/PropertiesReader.java:32]: Good catch. I was doing property overloading based on order of precedence then changed to load one of them. I forgot to reverse the order.

Fixed.

File: config-spring-property/src/main/java/org/stellar/anchor/server/config/payment/PropertyStellarxPaymentConfig.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+  private String name = "";
+  private boolean enabled = false;
+  private String horizonUrl;
```
[the remark this answers is not in the document]
Comment (lijamie98) [reply config-spring-property/src/main/java/org/stellar/anchor/server/config/payment/PropertyStellarxPaymentConfig.java:10]: The passphrase is already part of the `org.stellar.sdk.Network` class.

File: docs/anchor-platform.md (hunk 1 above)
Comment (lijamie98) [reply docs/anchor-platform.md:5]: Can be either absolute or relative depends on the file system. 

eg: `//home/user/.anchor/anchor-config.yaml` is absolute
eg: `./anchor/anchor-config.yaml` is relative.

File: platform/example.anchor-config.yaml (hunk 6 above)
Comment (lijamie98) [reply platform/example.anchor-config.yaml:25]: Changed.

File: platform/example.anchor-config.yaml (hunk 5 above)
Comment (lijamie98) [reply platform/example.anchor-config.yaml:8]: Deprecated documentation. Removed.

File: platform/example.anchor-config.yaml (hunk 7 above)
Comment (lijamie98) [reply platform/example.anchor-config.yaml:42]: If we allow all, we will use `clientAttributionDenyList`

File: platform/example.anchor-config.yaml (hunk 8 above)
Comment (lijamie98) [reply platform/example.anchor-config.yaml:28]: Changed to `horizonUrl` to be consistent.

File: platform/example.anchor-config.yaml (hunk 9 above)
Comment (lijamie98) [reply platform/example.anchor-config.yaml:61]: changed.

**Pull request merged**